rse-py-scripting.Rmd

```diff
@@ -103,42 +133,461 @@ optional arguments:
   -h, --help  show this help message and exit
 ```
 
-and issues errors when users give the program invalid arguments:
+As the `--help` hint shows,
+`argparse` can handle [long options][long-option-shell] as well as traditional single-letter options;
+the online help has examples of this and much more.
+
+## How can I configure common options for a program? {#rse-py-scripting-configure}
+
+A program that does exactly the same thing every time we run it
+isn't as useful as one that can work on different files
+or analyze data with different thresholds.
+Software of all kinds needs to be controlled;
+some things change more often than others,
+so we need a simple, uniform way to specify some options and leave others alone.
+
+The modern Unix convention is to provide multiple levels of configuration:
+
+1.  A system-wide configuration file for general settings.
+2.  A user-specific configuration file for personal preferences.
+3.  A job-specific file with settings for a specific run.
+4.  Command-line options to change things that commonly change.
+
+This is sometimes called [overlay configuration][overlay-configuration]
+because each level overrides the ones above it:
+the user's configuration file overrides the system settings,
+the job configuration overrides the user's defaults,
+and the command-line options overrides that.
+
+Programmers have invented many formats for configuration files,
+so please do not create your own.
+One possibility is to write the configuration as a Python data structure
+and then load it as if it was a library.
+This is clever,
+but it's hard for tools in other languages to process.
+Programers are also fond of [JSON][json],
+but that involves a lot of curly braces.
+A third option is the [Windows INI format][ini-format],
+which is laid out like this:
 
-```shell
-$ python script_template.py in.csv
+```text
+[section_1]
+key_1=value_1
+key_2=value_2
+
+[section_2]
+key_3=value_3
+key_4=value_4
+```
+
+INI files are simple to read and write,
+but the format is slowly falling out of use in favor of [YAML][yaml]
+(Appendix \@ref(rse-yaml)).
+Here's a sample YAML configuration file:
+
+```yaml
+# Standard settings for thesis.
+logfile: "/tmp/log.txt"
+quiet: false
+overwrite: false
+fonts:
+- Verdana
+- Serif
+```
+
+And here's a short Python program that reads and prints that configuration:
+
```

Comment:
Should this example yaml code be added to `script_template.py` so that we've got an example script that can be configured via command line options and/or a YAML file?

Reply:
Good catch - example enlarged.